include/stronk/utilities/dimensions.h

```diff
@@ -0,0 +1,175 @@
+                return DimensionsMerge<As...>::template merge<AccumulatedDimensionListT, Bs...>();
+            } else {
+                using new_merged_list = typename AccumulatedDimensionListT::template push_back_t<new_dim>;
+                return DimensionsMerge<As...>::template merge<new_merged_list, Bs...>();
+            }
+        } else if constexpr (t_before_o) {
+            using new_merged_list = typename AccumulatedDimensionListT::template push_back_t<A>;
+            return DimensionsMerge<As...>::template merge<new_merged_list, B, Bs...>();
+        } else {
+            using new_merged_list = typename AccumulatedDimensionListT::template push_back_t<B>;
+            return DimensionsMerge<A, As...>::template merge<new_merged_list, Bs...>();
+        }
+    }
+
+    template<dimension_like... OtherTs>
+    using merge_t = decltype(merge<EmptyDimensions, OtherTs...>());
+};
+
+// Use create_dimensions_t to instantiate this type
+template<dimension_like... Ts>
+struct Dimensions
+{
+    [[nodiscard]] constexpr static auto size() -> std::size_t { return sizeof...(Ts); }
+
+    [[nodiscard]] constexpr static auto empty() -> bool { return size() == 0; }
+
+    [[nodiscard]] constexpr static auto is_pure() -> bool { return size() == 1 && first_t::rank == 1; }
+
+    using first_t = stronk_details::variadic::first_type_of_t<Ts..., Dimension<void, 0>>;  // void if empty
+
+    template<dimension_like NewDimT>
+    using push_back_t = Dimensions<Ts..., NewDimT>;
+
+    template<dimension_like... NewDimTs>
+    using concat_t = Dimensions<Ts..., NewDimTs...>;
+
+    using negate_t = Dimensions<typename Ts::negate_t...>;
+
+    template<dimension_like... Bs>
+    static auto multiply(Dimensions<Bs...>)
+    {
+        return DimensionsMerge<Ts...>::template merge<Dimensions<>, Bs...>();
+    }
+    template<typename OtherDimensionsT>
+    using multiply_t = decltype(multiply(OtherDimensionsT {}));
+
+    template<dimension_like... Bs>
+    static auto divide(Dimensions<Bs...>)
+    {
+        using negated_b = typename Dimensions<Bs...>::negate_t;
+        return multiply(negated_b {});
+    }
+    template<typename OtherDimensionsT>
+    using divide_t = decltype(divide(OtherDimensionsT {}));
+};
+
+template<typename... ExistingDimTs>
+auto create(Dimensions<ExistingDimTs...> dim)
+{
+    return dim;
+}
+
+template<typename... ExistingDimTs, dimension_like Dim, dimension_like... DimTs>
+auto create(Dimensions<ExistingDimTs...> dims, Dim, DimTs... rest)
+{
+    return create(dims.multiply(details::Dimensions<Dim> {}), rest...);
+}
+
+}  // namespace details
+
+using EmptyDimensions = details::EmptyDimensions;
+template<typename T>
+concept dimensions_like = requires(T) {
+    typename T::first_t;
+    T::size();
+    T::empty();
+    typename T::negate_t;
+    typename T::template multiply_t<EmptyDimensions>;
+    typename T::template divide_t<EmptyDimensions>;
+};
```

Comment:
If you define this earlier, you can use it in the type parameters on e.g. Dimensions::divide_t

Reply:
we tried and it make the definition recursive